### PR TITLE
IFP128

### LINKED ISSUE #113: `IFP128`
### Motivation

Fixed point integers currently do not have a signed 128-bit integer type.

### PR DESCRIPTION
**Pull request by rostyslavtyshko**
## Type of change

- New feature
Added `IFP128` - 128-bit signed fixed point number

## Changes

The following changes have been made:

-Added `IFP128` - 128-bit signed fixed point number

## Notes

Closes #113 

## Related Issues

Closes #\<issue number\>

### CODE REVIEW

**Review by nfurfaro — DISMISSED** [review]

File: libs/fixed_point/README.md [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 # Overview
 
-The Fixed Point Number library provides a library to use fixed-point numbers in Sway. It has 3 distinct unsigned types: `UFP32`, `UFP64` and `UFP128` as well as a signed type - `IFP64`. This type is stack allocated.
+The Fixed Point Number library provides a library to use fixed-point numbers in Sway. It has 3 distinct unsigned types: `UFP32`, `UFP64` and `UFP128` as well as 2 signed types - `IFP64` and `IFP128`. Thess types are stack allocated.
```
Comment (bitzoic) [inline libs/fixed_point/README.md:10]:
```suggestion
The Fixed Point Number library provides a library to use fixed-point numbers in Sway. It has 3 distinct unsigned types: `UFP32`, `UFP64`, and `UFP128` as well as 2 signed types - `IFP64` and `IFP128`. These types are stack allocated.

```

File: libs/fixed_point/src/ifp128.sw [hunk 2]
```diff
@@ -0,0 +1,306 @@
+library ifp128;
+// A wrapper library around the  type for mathematical functions operating with signed 128-bit fixed point numbers.
```
Comment (bitzoic) [inline libs/fixed_point/src/ifp128.sw:2]:
```suggestion
// A wrapper library around the type for mathematical functions operating with signed 128-bit fixed point numbers.
```

**Review by bitzoic — CHANGES_REQUESTED** [review]
Minor comments but looks good!

File: libs/fixed_point/src/ifp128.sw [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+    fn into(self) -> UFP64 {
+        self.underlying
+    }
+}
+
+impl IFP128 {
+    /// The size of this type in bits.
+    pub fn bits() -> u32 {
+        72
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply libs/fixed_point/src/ifp128.sw:28]: 64 bits for `UFP64` + 8 bits for `bool`

File: libs/fixed_point/README.md (hunk 1 above)
Comment (rostyslavtyshko) [reply libs/fixed_point/README.md]: Fixed

File: libs/fixed_point/src/ifp128.sw [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        }
+        Self {
+            underlying: underlying,
+            non_negative: self.non_negative,
+        }
+    }
+}
+
+impl IFP128 {
+    /// Returns the nearest integer to `self`. Round half-way cases away from
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply libs/fixed_point/src/ifp128.sw:240]: Fixed

File: libs/fixed_point/src/ifp128.sw [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+            underlying: UFP64::min(),
+            non_negative: false,
+        }
+    }
+
+    pub fn zero() -> Self {
+        Self::from(UFP64::zero())
+    }
+
+    fn sign_reverse(self) -> Self {
```
[the remark this answers is not in the document]
Comment (rostyslavtyshko) [reply libs/fixed_point/src/ifp128.sw]: Made public

**Review by bitzoic — APPROVED** [review]
:shipit:

**Pull request merged**